signing/keyExtractor.go

```diff
@@ -0,0 +1,62 @@
+package signing
+
+import (
+	"github.com/spacemeshos/ed25519"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+)
+
+type edExtractorOption struct {
+	prefix []byte
+}
+
+// ExtractorOptionFunc to modify verifier.
+type ExtractorOptionFunc func(*edExtractorOption) error
+
+// WithExtractorPrefix sets the prefix used by PubKeyExtractor. This usually is the Network ID.
+func WithExtractorPrefix(prefix []byte) ExtractorOptionFunc {
+	return func(opts *edExtractorOption) error {
+		opts.prefix = prefix
+		return nil
+	}
+}
+
+// PubKeyExtractor extracts public keys from signatures.
+type PubKeyExtractor struct {
+	prefix []byte
+}
+
+// NewPubKeyExtractor returns a new PubKeyExtractor.
+func NewPubKeyExtractor(opts ...ExtractorOptionFunc) (*PubKeyExtractor, error) {
+	cfg := &edExtractorOption{}
+	for _, opt := range opts {
+		if err := opt(cfg); err != nil {
+			return nil, err
+		}
+	}
+	extractor := &PubKeyExtractor{
+		prefix: cfg.prefix,
+	}
+	return extractor, nil
+}
+
+// Extract public key from a signature.
+func (e PubKeyExtractor) Extract(m, sig []byte) (*PublicKey, error) {
+	msg := make([]byte, len(m)+len(e.prefix))
+	copy(msg, e.prefix)
+	copy(msg[len(e.prefix):], m)
+	pub, err := ed25519.ExtractPublicKey(msg, sig)
+	if err != nil {
+		return nil, err
+	}
+	return &PublicKey{PublicKey: pub}, nil
+}
+
+// ExtractNodeID from a signature.
+func (e PubKeyExtractor) ExtractNodeID(m, sig []byte) (types.NodeID, error) {
+	msg := make([]byte, len(m)+len(e.prefix))
+	copy(msg, e.prefix)
+	copy(msg[len(e.prefix):], m)
+	pub, err := ed25519.ExtractPublicKey(msg, sig)
```

Comment:
would this cause error if pub is `nil`?

i think it's better to reuse code from Extract()

Reply:
if `pub` is `nil` `types.BytesToNodeID` will return an empty `[32]byte` which is the base for the `NodeId` type. This is safe to do and will also result in a `NodeID` that equals `types.EmptyNodeID`.

re-using `Extract` here requires me to check for an error twice (once in `Extract` and once in `ExtractNodeID` or I'll get a panic due to a `nil` pointer dereference in the case of an error:

```golang
func (e PubKeyExtractor) ExtractNodeID(m, sig []byte) (types.NodeID, error) {
	pub, err := e.Extract(m, sig)
	if err != nil {
		return types.EmptyNodeID, err
	}
	return types.BytesToNodeID(pub.PublicKey), nil
}
```

Would you prefer this over the current version?

Reply:
checking error twice is trivial. using the same code to extract is more important to me.

Reply:
Updated code for ExtractNodeID